### ISSUE TITLE
Why yolo giving bad result on the test dataset over increasing the epochs.

### ISSUE THREAD

**Issue by Amalu97** [ultralytics/ultralytics#22496]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I am using the yolo nano version for the fire detection using the FlameVisions dataset for the object detcetion, https://www.kaggle.com/datasets/anamibnjafar0/flamevision/data . I used pretrained weight of the yolo while strated the training and I did the experiment for 100, 300 and 500 epochs.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Amalu97, thank you for your interest in Ultralytics 🚀! This is an automated response—an Ultralytics engineer will also review your issue and assist soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible so we can help diagnose performance over epochs:
-Exact model and weights used (e.g., YOLO11n, pretrained=True)  
-Full training/validation command(s) and any custom hyperparameters  
-Training logs and metrics (mAP, precision/recall) across epochs  
-Key plots: `results.png`, PR/F1 curves, confusion matrix, sample predictions  
-Dataset details: data.yaml, class list, train/val/test split sizes, example images  
-Environment details: OS, Python, PyTorch, CUDA, `ultralytics` version  
Also verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) 👍

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
Can you post the `results.jpg` and `labels.jpg` file from `runs` folder?

**Comment by Amalu97** [comment]
[results.pdf](https://github.com/user-attachments/files/23175577/results.pdf) this contain the training and testing results

**Comment by glenn-jocher** [comment]
Thanks for the PDF; to help diagnose the test-dropoff (likely overfitting on a small dataset), please attach the images `runs/detect/train/results.png` (or `.jpg`), `runs/detect/train/confusion_matrix.png`, and your dataset `labels.jpg`, plus paste the text of `runs/detect/train/results.csv` (epochs/mAP/precision/recall) and your `data.yaml` so we can verify disjoint splits; also share your environment via:

```bash
yolo checks
```

To confirm the issue on your true test set, run:

```bash
yolo mode=val model=runs/detect/train/weights/best.pt data=path/to/data.yaml split=test plots=True
```

If confirmed, try early stopping and shorter training with YOLO11n (nano) to avoid overfitting; see the guidance in [Troubleshooting Common YOLO Issues](https://docs.ultralytics.com/guides/yolo-common-issues/) and the overview on [preventing overfitting in computer vision](https://www.ultralytics.com/blog/what-is-overfitting-in-computer-vision-how-to-prevent-it).

**Comment by Y-T-G** [comment]
@Amalu97 You can try running val without all the extra arguments. Just use `model.val(data="data.yaml", split="test")`. Remove all th other arguments. Those metrics mean that the test dataset has issues. You should also check the `val_batch_labels.jpg` files inside runs folder to see if the labels are correct or missing.

**Comment by Amalu97** [comment]
Here I am attaching the document with all the requested result. I would also like to know while start the training its freezing the last block of the model model.23.dfl.conv.weight - the head of the model. @Y-T-G @glenn-jocher 

[All_results.pdf](https://github.com/user-attachments/files/23218620/All_results.pdf)

**Comment by glenn-jocher** [comment]
Thanks for the bundle. On “freezing `model.23.dfl.conv.weight`”: YOLO doesn’t freeze the detection head by default—this only happens if `freeze` is set. Please share your exact training command and attach `runs/detect/train/args.yaml` so we can confirm. You can also verify locally with:

```python
from ultralytics import YOLO
m = YOLO("runs/detect/train/weights/last.pt")
for n, p in m.model.named_parameters():
    if "dfl" in n:
        print(n, p.requires_grad)  # should be True if not frozen
```

If any show `False`, remove custom freeze lists or run with `freeze=0` (see the `freeze` argument in [Model Training with Ultralytics YOLO](https://docs.ultralytics.com/modes/train/)).  

For the test drop-off, please paste the text of `runs/detect/train/results.csv` and `yolo checks` (no PDFs), and attach `val_batch_labels.jpg`; if you fix any labels, delete `labels/*.cache` before re-validating. To review metrics/plots on your held-out set, use validation as described in [Model Validation with Ultralytics YOLO](https://docs.ultralytics.com/modes/val/).

**Comment by Amalu97** [comment]
@glenn-jocher Thank you for your replay. Regarding freezing of model.23.dfl.conv.weight, what my question was, eventhough I did not gave any freez value by default befor showing the epochs of training it showing the message like freezing the model.23.dfl.conv.weight layer, i do not know does it will become unfreez during the training process. Below I am attaching the python file with the program and results .  If you open it in jupyterlab or Pycharm you can see the already ran epoch results.  Also I am attaching the result.csv file for your reference. on the pdf i share above at the end of it you can see the different val_batch_labels.jpg corresponding to batch 0,1 and 2 and its predictions. I will attach the args.yaml file later today.

[Exp22.ipynb](https://github.com/user-attachments/files/23222228/Exp22.ipynb)

[results.csv](https://github.com/user-attachments/files/23222233/results.csv)

**Comment by Y-T-G** [comment]
Is this `val_batch` image before or after running on the test set? It seems like from `val` set and not `test` set. I want to see the ones from the test set. Did you try running without all the extra arguments like I mentioned?

**Comment by Amalu97** [comment]
The images are from test dataset, because of we are running it in val mode its create the name like that.. In the file I shared above you can see the code I used for run inference on test dataset @Y-T-G

**Comment by Amalu97** [comment]
Below I am attaching the arg.yaml file @glenn-jocher 

[args.yaml](https://github.com/user-attachments/files/23234199/args.yaml)

**Comment by Y-T-G** [comment]
> In the file I shared above you can see the code I used for run inference on test dataset

It still uses a lot of extra arguments. You should try running with only `data` and `split` argument

**Comment by Amalu97** [comment]
> > In the file I shared above you can see the code I used for run inference on test dataset
> 
> It still uses a lot of extra arguments. You should try running with only `data` and `split` argument

I tried this way but the resulting values are same

**Comment by Y-T-G** [comment]
Form your test images, if your training set is also like that, then you trained the model on similar images just captured from slightly different angles which is bad because that will make the model overfit. You need to train on diverse images and not just similar images for a good model.

**Comment by Amalu97** [comment]
> Thanks for the bundle. On “freezing `model.23.dfl.conv.weight`”: YOLO doesn’t freeze the detection head by default—this only happens if `freeze` is set. Please share your exact training command and attach `runs/detect/train/args.yaml` so we can confirm. You can also verify locally with:
> 
> from ultralytics import YOLO
> m = YOLO("runs/detect/train/weights/last.pt")
> for n, p in m.model.named_parameters():
>     if "dfl" in n:
>         print(n, p.requires_grad)  # should be True if not frozen
> If any show `False`, remove custom freeze lists or run with `freeze=0` (see the `freeze` argument in [Model Training with Ultralytics YOLO](https://docs.ultralytics.com/modes/train/)).
> 
> For the test drop-off, please paste the text of `runs/detect/train/results.csv` and `yolo checks` (no PDFs), and attach `val_batch_labels.jpg`; if you fix any labels, delete `labels/*.cache` before re-validating. To review metrics/plots on your held-out set, use validation as described in [Model Validation with Ultralytics YOLO](https://docs.ultralytics.com/modes/val/).

I tried to run this code, then I could see that it returning model.23.dfl.conv.weight as false, also I tried with freez =0 at that time also model.23.dfl.conv.weight is freezing while start training. It would be great if you can share your insight on this problem based on the requested docs I uploded. Thanks in advance @glenn-jocher

**Comment by Y-T-G** [comment]
That layer is always frozen

**Comment by Amalu97** [comment]
I have one more doubt, its regarding the confusion matrix, for example in fire only detection, for background detected ad background (TN) why its display in white colour as in the attached photo. Is it possible to change that column color to another colour? @glenn-jocher 

![Image](https://github.com/user-attachments/assets/7d468144-3f02-4334-aa7d-609bd6d546be)

**Comment by Amalu97** [comment]
Also I would like to know does the background count alway will be zero in the confusion_matrix? @glenn-jocher

**Comment by glenn-jocher** [comment]
@Amalu97 On `model.23.dfl.conv.weight`: that layer is created with `requires_grad=False` by design in YOLO, because it just holds fixed bin positions for the DFL box-regression step; it stays “frozen” for all epochs and is not controlled by the `freeze` argument, but the rest of the head still trains normally, so you can safely ignore that message as a cause of your test drop-off.  

For the confusion matrix, the extra “background” row/column aggregate false negatives (missed objects) and false positives (detections where there is no ground-truth object); the bottom-right background cell (true negatives) is effectively undefined for detection and is always stored as zero, which maps to white with the default colormap, so seeing that cell white and its count at 0 is expected. If you’d like different coloring or behavior you can load `results.confusion_matrix.matrix` from your run and re-plot it with your own matplotlib colormap as described in the confusion-matrix overview at [Confusion Matrix Explained](https://www.ultralytics.com/glossary/confusion-matrix).
